tests/simulation/testSphereParticleGenerator.cpp

```diff
@@ -0,0 +1,234 @@
+    SphereParticleCluster cluster (
+        origin,
+        15,
+        3,
+        spacing,
+        mass,
+        initialVelocity,
+        meanVelocity,
+        dimensions);
+    generator.registerCluster(std::make_unique<SphereParticleCluster>(cluster));
+    generator.generateClusters();
+
+    for (auto it = container.beginPairs(); it != container.endPairs(); ++it) {
+        std::pair<Particle&, Particle&> pt_pair = *it;
+
+        std::array<double, 3> pos1 = pt_pair.first.getX(), pos2 = pt_pair.second.getX();
+        double dist = ArrayUtils::L2Norm(pos1 - pos2);
+
+        EXPECT_GT(dist, 0);
+    }
+}
+
+// Test if the cluster spans a sphere i.e. if the distance to the origin is less or equal to the radius
+TEST_F(SphereParticleGeneratorTest, ClusterSpansSphere)
+{
+    size_t radius = 15;
+
+    ParticleGenerator generator { this->container };
+    SphereParticleCluster cluster (
+        origin,
+        radius,
+        3,
+        spacing,
+        mass,
+        initialVelocity,
+        meanVelocity,
+        dimensions);
+    generator.registerCluster(std::make_unique<SphereParticleCluster>(cluster));
+    generator.generateClusters();
+
+    for (auto& p : container) {
+        double dist = ArrayUtils::L2Norm(p.getX() - origin);
+        EXPECT_LE(dist, radius * spacing);
+    }
+}
+
+// Test if the 2D sphere is stable i.e. the particles do not move more than 5% out of the radius during simulation
+TEST_F(SphereParticleGeneratorTest, Cluster2DStable)
+{
+    size_t radius = 10;
+    double delta_t = 0.0014;
+    double end_time = delta_t * 1000;
+
+    double tolerance = 1.05;
+
+    ParticleGenerator generator { container };
+    SphereParticleCluster cluster (
+        origin,
+        radius,
+        2,
+        spacing,
+        mass,
+        initialVelocity,
+        meanVelocity,
+        2);
+    generator.registerCluster(std::make_unique<SphereParticleCluster>(cluster));
+    generator.generateClusters();
+
+    // Initialize reader
+    EmptyFileReader fileReader("");
+    // Initialize writer
+    EmptyFileWriter writer;
+
+    auto readerPointer = std::make_unique<EmptyFileReader>(fileReader);
+    auto writerPointer = std::make_unique<EmptyFileWriter>(writer);
+
+    // Intialize physics strategy
+    PhysicsStrategy strat { location_stroemer_verlet,
+                            velocity_stroemer_verlet,
+                            force_lennard_jones };
+
+
+    // Setup simulation
+    LennardJonesSimulation sim (0, delta_t, end_time, container, strat, std::move(writerPointer), std::move(readerPointer), 5, 1 );
+
+    double time = 0;
+    int iteration = 0;
+    while (time < end_time) {
+
+        if (iteration++ % 10 == 0) {
+            // Check if distance is leq than 105% of the radius
+            for (auto& p : container) {
+                double dist = ArrayUtils::L2Norm(p.getX() - origin);
+                EXPECT_LE(dist, radius * spacing * tolerance);
+            }
+        }
```

Comment:
Hier schaust du ob die Distanz innerhalb 5% zusätzlich liegt, aber um zu testen ob die Partikel innerhalb der allgemeinen 5% bleiben, müssen wir auch von unten vergleichen (solange ich das richtig verstanden habe):
```suggestion
        if (iteration++ % 10 == 0) {
            // Check if distance is leq than 105% of the radius
            for (auto& p : container) {
                double dist = ArrayUtils::L2Norm(p.getX() - origin);
                EXPECT_LE(dist, radius * spacing * tolerance) << "Particle breached upper tolerance!";
                EXPECT_GE(dist, radius * spacing * 0.95) << "Particle breached lower tolerance!";
            }
        }
```
Vielleicht noch die 0.95 in eine Variable packen, oder `tolerance` auf 0.05 setzen und dann mit `(1+tolerance)` und `(1-tolerance)` rechnen.

Reply:
Das Problem ist, dass wenn man es so testet, dann hat man quasi nur eine Hülle
Also quasi eine Schicht, in der alle Teilchen sein dürfen und die hat eine Breite von 10% des gesamt Radius

Unsere Spheres sind aber solide, heißt die gehen bis ins Zentrum (origin)

<img width="257" alt="image" src="https://github.com/noahpy/MolSim-SS24/assets/56398851/ad317519-1ba3-4fff-9c36-9d6f1085fd44">

Mit dem Test wollte ich nur schauen, ob die Kugel an sich stabil ist und nicht einfach "explodiert"
Macht das Sinn?

Reply:
Ah ja ok ergibt schon Sinn :)